extension/chrome/content/download.js

```diff
@@ -83,18 +85,38 @@ Downloader = {
 
       // If target doesn't exist, create it
       if (!this.target.exists()) {
-        this.target.create(0x00, 0644);
+        this.target.create(Ci.nsIFile.NORMAL_FILE_TYPE, parseInt("0666", 8));
       }
 
-      this._persist = Cc["@mozilla.org/embedding/browser/nsWebBrowserPersist;1"].
-                      createInstance(Ci.nsIWebBrowserPersist);
-      this._persist.persistFlags = Ci.nsIWebBrowserPersist.PERSIST_FLAGS_REPLACE_EXISTING_FILES |
-                                   Ci.nsIWebBrowserPersist.PERSIST_FLAGS_FROM_CACHE |
-                                   Ci.nsIWebBrowserPersist.PERSIST_FLAGS_CLEANUP_ON_FAILURE;
-
-      // Save URL as target
-      this._persist.progressListener = new DownloaderListener();
-      this._persist.saveURI(uri, null, null, null, null, this.target, null);
+      Downloads.createDownload({'source': uri, 'target': this.target}).then((aDownload) => {
+        var maxMB = -1;
+        var currentMB = -1;
+        aDownload.onchange = function () {
+          if (aDownload.succeeded) {
+            var bytes = aDownload.hasProgress ?
+                        aDownload.totalBytes : aDownload.currentBytes;
+            maxMB = bytes / 1024 / 1024;
```

Comment:
If a download is marked as succeeded all bytes have been downloaded. So I don't think that we have to make this check and can directly use totalBytes. Also instead of dividing each time by 1024 use a constant for it.